File: workers/go/cmd/http-worker/handler.go
package main

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"

	cw "github.com/kinarix/conduit/workers/go/conduitworker"
)

// HTTPHandler binds one HandlerConfig to one topic and runs the side
// effect on each task. Mirrors workers/rust/crates/http-worker/src/handler.rs.
type HTTPHandler struct {
	topic  string
	cfg    HandlerConfig
	client *http.Client
}

func NewHTTPHandler(topic string, cfg HandlerConfig) *HTTPHandler {
	return &HTTPHandler{
		topic:  topic,
		cfg:    cfg,
		client: &http.Client{Timeout: cfg.Timeout()},
	}
}

// Handle implements cw.HandlerFunc.
func (h *HTTPHandler) Handle(ctx context.Context, task *cw.ExternalTask) (*cw.Result, error) {
	vars := decodeVariables(task)
	taskID := task.ID

	url := render(h.cfg.URLTemplate, vars, taskID)

	var body io.Reader
	if h.cfg.RequestTemplate != nil {
		rendered := renderJSON(h.cfg.RequestTemplate, vars, taskID)
		buf, err := json.Marshal(rendered)
		if err != nil {
			return nil, fmt.Errorf("marshal request body: %w", err)
		}
		body = bytes.NewReader(buf)
	}

	req, err := http.NewRequestWithContext(ctx, h.cfg.Method, url, body)
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}

	for k, v := range h.cfg.Headers {
		req.Header.Set(k, render(v, vars, taskID))
	}
	if name, value, err := h.authHeader(); err != nil {
		return nil, err
	} else if name != "" {
		req.Header.Set(name, value)
	}
	if h.cfg.Idempotency.IdempotencyEnabled() {
		key := render(h.cfg.Idempotency.KeyTemplate, vars, taskID)
		req.Header.Set(h.cfg.Idempotency.Header, key)
	}
	if body != nil && req.Header.Get("Content-Type") == "" {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := h.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("transport: %w", err)
	}
	defer resp.Body.Close()
	bodyBytes, _ := io.ReadAll(resp.Body)

	switch {
	case resp.StatusCode >= 200 && resp.StatusCode < 300:
		return h.successResult(bodyBytes), nil
	case resp.StatusCode >= 400 && resp.StatusCode < 500 && h.cfg.BpmnErrorOn4xx != "":
		return cw.BpmnError(
			h.cfg.BpmnErrorOn4xx,
			fmt.Sprintf("HTTP %d: %s", resp.StatusCode, truncate(string(bodyBytes), 256)),
			cw.VarLong("http_status", int64(resp.StatusCode)),
			cw.VarString("http_body", string(bodyBytes)),
		), nil
	default:
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, truncate(string(bodyBytes), 256))
	}
}

func (h *HTTPHandler) successResult(bodyBytes []byte) *cw.Result {
	var bodyJSON any
	_ = json.Unmarshal(bodyBytes, &bodyJSON)

	out := make([]cw.Variable, 0, len(h.cfg.ResponseMapping))
	for varName, path := range h.cfg.ResponseMapping {
		v, ok := jsonpath(path, bodyJSON)
		if !ok {
			out = append(out, cw.VarNull(varName))
			continue
		}
		varJSON, err := cw.VarJSON(varName, v)
		if err != nil {
			out = append(out, cw.VarNull(varName))
			continue
		}
		out = append(out, varJSON)
	}
	return cw.Complete(out...)
}

// decodeVariables unmarshals each wire variable's raw JSON into a Go value
// so render() can substitute templates against typed values.
func decodeVariables(task *cw.ExternalTask) map[string]any {
	out := make(map[string]any, len(task.Variables))
	for _, v := range task.Variables {
		var decoded any
		if err := json.Unmarshal(v.Value, &decoded); err != nil {
			out[v.Name] = nil
			continue
		}
		out[v.Name] = decoded
	}
	return out
}

func (h *HTTPHandler) authHeader() (name, value string, err error) {
	if h.cfg.Auth == nil {
		return "", "", nil
	}
	switch h.cfg.Auth.Type {
	case "bearer":
		token := os.Getenv(h.cfg.Auth.TokenEnv)
		if token == "" {
			return "", "", fmt.Errorf("env var %s (auth.token_env) not set", h.cfg.Auth.TokenEnv)
		}
		return "Authorization", "Bearer " + token, nil
	case "basic":
		user := os.Getenv(h.cfg.Auth.UserEnv)
		pass := os.Getenv(h.cfg.Auth.PasswordEnv)
		if user == "" || pass == "" {
			return "", "", fmt.Errorf(
				"env vars %s/%s (auth.user_env/password_env) not set",
				h.cfg.Auth.UserEnv, h.cfg.Auth.PasswordEnv,
			)
		}
		creds := base64.StdEncoding.EncodeToString([]byte(user + ":" + pass))
		return "Authorization", "Basic " + creds, nil
	default:
		return "", "", fmt.Errorf("unsupported auth.type: %q", h.cfg.Auth.Type)
	}
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	// Walk back to a UTF-8 boundary.
	for i := max; i > 0; i-- {
		if !strings.HasPrefix(s[i:], "¿½") && (s[i]&0xc0) != 0x80 {
			return s[:i] + "…"
		}
	}
	return s[:max] + "…"
}

File: workers/rust/crates/http-worker/src/handler.rs
use async_trait::async_trait;
use conduit_worker::{ExternalTask, Handler, HandlerError, HandlerResult, Variable};
use reqwest::{header, Client as HttpClient, Method};
use std::str::FromStr;

use crate::config::{AuthConfig, HandlerConfig};
use crate::render::{jsonpath, render, render_json};

/// One handler instance, bound to one topic.
pub struct HttpHandler {
    topic: String,
    config: HandlerConfig,
    http: HttpClient,
    auth_header: Option<(header::HeaderName, String)>,
}

impl HttpHandler {
    pub fn new(topic: String, config: HandlerConfig) -> Result<Self, BuildError> {
        let http = HttpClient::builder()
            .timeout(config.timeout())
            .build()
            .map_err(BuildError::from)?;
        let auth_header = resolve_auth_header(config.auth.as_ref())?;
        Ok(Self {
            topic,
            config,
            http,
            auth_header,
        })
    }
}

fn resolve_auth_header(
    auth: Option<&AuthConfig>,
) -> Result<Option<(header::HeaderName, String)>, BuildError> {
    let Some(auth) = auth else {
        return Ok(None);
    };
    match auth {
        AuthConfig::Bearer { token_env } => {
            let token = std::env::var(token_env)
                .map_err(|_| BuildError::Env(format!("{token_env} (auth.token_env)")))?;
            Ok(Some((header::AUTHORIZATION, format!("Bearer {token}"))))
        }
        AuthConfig::Basic {
            user_env,
            password_env,
        } => {
            let user = std::env::var(user_env)
                .map_err(|_| BuildError::Env(format!("{user_env} (auth.user_env)")))?;
            let pass = std::env::var(password_env)
                .map_err(|_| BuildError::Env(format!("{password_env} (auth.password_env)")))?;
            let creds = base64_pretend::encode(format!("{user}:{pass}"));
            Ok(Some((header::AUTHORIZATION, format!("Basic {creds}"))))
        }
    }
}

#[async_trait]
impl Handler for HttpHandler {
    fn topic(&self) -> &str {
        &self.topic
    }

    async fn handle(&self, task: &ExternalTask) -> Result<HandlerResult, HandlerError> {
        let vars = task.variable_map();
        let task_id_str = task.id.to_string();

        let url = render(&self.config.url_template, &vars, &task_id_str);
        let method = Method::from_str(&self.config.method.to_uppercase()).map_err(|_| {
            HandlerError::new(format!("invalid HTTP method: {}", self.config.method))
        })?;

        let mut req = self.http.request(method, &url);

        for (k, v) in &self.config.headers {
            let v = render(v, &vars, &task_id_str);
            req = req.header(k.as_str(), v);
        }
        if let Some((name, value)) = &self.auth_header {
            req = req.header(name, value);
        }
        if self.config.idempotency.enabled {
            let key = render(&self.config.idempotency.key_template, &vars, &task_id_str);
            req = req.header(self.config.idempotency.header.as_str(), key);
        }

        if let Some(body) = &self.config.request_template {
            let rendered = render_json(body, &vars, &task_id_str);
            req = req.json(&rendered);
        }

        let resp = req
            .send()
            .await
            .map_err(|e| HandlerError::new(format!("transport: {e}")))?;
        let status = resp.status();
        let body_text = resp.text().await.unwrap_or_default();

        if status.is_success() {
            let body_json: serde_json::Value =
                serde_json::from_str(&body_text).unwrap_or(serde_json::Value::Null);
            let mut out_vars = Vec::with_capacity(self.config.response_mapping.len());
            for (var_name, path) in &self.config.response_mapping {
                let value = jsonpath(path, &body_json)
                    .cloned()
                    .unwrap_or(serde_json::Value::Null);
                out_vars.push(Variable::json(var_name.clone(), value));
            }
            return Ok(HandlerResult::Complete {
                variables: out_vars,
            });
        }

        if status.is_client_error() {
            if let Some(code) = &self.config.bpmn_error_on_4xx {
                return Ok(HandlerResult::BpmnError {
                    code: code.clone(),
                    message: format!("HTTP {status}: {}", truncate(&body_text, 256)),
                    variables: vec![
                        Variable::long("http_status", status.as_u16() as i64),
                        Variable::string("http_body", body_text),
                    ],
                });
            }
        }

        Err(HandlerError::new(format!(
            "HTTP {status}: {}",
            truncate(&body_text, 256)
        )))
    }
}

#[derive(Debug, thiserror::Error)]
pub enum BuildError {
    #[error("reqwest: {0}")]
    Reqwest(#[from] reqwest::Error),
    #[error("env var not set: {0}")]
    Env(String),
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() <= max {
        return s.to_string();
    }
    let mut boundary = max;
    while boundary > 0 && !s.is_char_boundary(boundary) {
        boundary -= 1;
    }
    format!("{}…", &s[..boundary])
}

// Tiny base64 stand-in so we don't pull a full dep just for Basic auth.
mod base64_pretend {
    const ALPHABET: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    pub fn encode(input: impl AsRef<[u8]>) -> String {
        let bytes = input.as_ref();
        let mut out = String::with_capacity((bytes.len().div_ceil(3)) * 4);
        for chunk in bytes.chunks(3) {
            let b0 = chunk[0];
            let b1 = chunk.get(1).copied().unwrap_or(0);
            let b2 = chunk.get(2).copied().unwrap_or(0);
            out.push(ALPHABET[(b0 >> 2) as usize] as char);
            out.push(ALPHABET[(((b0 & 0x03) << 4) | (b1 >> 4)) as usize] as char);
            if chunk.len() > 1 {
                out.push(ALPHABET[(((b1 & 0x0F) << 2) | (b2 >> 6)) as usize] as char);
            } else {
                out.push('=');
            }
            if chunk.len() > 2 {
                out.push(ALPHABET[(b2 & 0x3F) as usize] as char);
            } else {
                out.push('=');
            }
        }
        out
    }

    #[test]
    fn encodes_known_pairs() {
        assert_eq!(encode("user:pass"), "dXNlcjpwYXNz");
        assert_eq!(encode(""), "");
        assert_eq!(encode("a"), "YQ==");
        assert_eq!(encode("ab"), "YWI=");
        assert_eq!(encode("abc"), "YWJj");
    }
}
